### ISSUE TITLE
Principled `Map#{...}`/`Set#{...}`/`[...`

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#4961]
Related to #4760 I was thinking about `Map#{...}`, `Set#{...}`, `[...]` and how to make it more principled. My motivation came from the desire to have `List#{...}`.

My idea is to have a `FromList` type class which the compiler will desuger `Type#{a, b, c}` into (`FromList.(a :: b :: c :: Nil)`). This works for `Set`/`Map`/`List` but not array and regioned stuff.

For this we could have the `RegionedFromList` type class, which the compiler will use for `Type#{a, b, c} @ r` (`RegionedFromList.(a :: b :: c :: Nil, r)`).

**Comment by mlutze** [comment]
`Map` has a different arity from `List` and `Set` so it doesn't directly work there. I see two options:

1. Another type class for Binary type constructors.
2. Associated type `(k, v)` as the list element

**Comment by JonathanStarup** [comment]
Yeah I also considered that after posting. hmm 🤔

What about a low tech, desugaring solution. `Someting#{...} @ r` just becomes `Something.fromList(r, ...)`.

**Comment by magnus-madsen** [comment]
If you insist, I am OK with adding `List#{1, 2, 3}`.

**Comment by JonathanStarup** [comment]
I've implemented `List#{...}` but I still believe that desugaring `MyOwnThing#{...}` to `MyOwnThing.fromList(List#{...})` would be cool

**Comment by JonathanStarup** [comment]
Wait, This is just reinventing scalas `apply` trick with extra steps. Should we add `NS(x, y, z)` as sugar for `NS.apply(x, y, z)`? Then we have `Set(...)` `Map(...)` etc for free

**Comment by JonathanStarup** [comment]
@magnus-madsen I think we have the power to do this and should do it
Pros: it simplifies the AST to one base literal (either List or Vector probably) and it makes the literal creation extensible
First: i propose a new syntax since `#{}` is datalog world. Lets just show examples with `|...|` and assume that vector is the base literal
- `|...|` is desugared `FromVector.fromVector($Vector$(...))` (creating an unknown immutable collection)
- `|...| @ rc` is desugared `FromVectorMut.fromVector(rc, $Vector$(...))` (creating an unknown mutable collection)
- `Type|...|` is desugared `(FromVector.fromVector($Vector$(...)): Type[_, _, ...])` (desugaring requires kind of Type)
- `Type|...| @ rc` is desugared `(FromVectorMut.fromVector(rc, $Vector$(...)): Type[_, _, ...])` (desugaring requires the Kind of Type)

**Comment by magnus-madsen** [comment]
> [@magnus-madsen](https://github.com/magnus-madsen) I think we have the power to do this and should do it Pros: it simplifies the AST to one base literal (either List or Vector probably) and it makes the literal creation extensible First: i propose a new syntax since `#{}` is datalog world. Lets just show examples with `|...|` and assume that vector is the base literal
> 
>     * `|...|` is desugared `FromVector.fromVector($Vector$(...))` (creating an unknown immutable collection)
> 
>     * `|...| @ rc` is desugared `FromVectorMut.fromVector(rc, $Vector$(...))` (creating an unknown mutable collection)
> 
>     * `Type|...|` is desugared `(FromVector.fromVector($Vector$(...)): Type[_, _, ...])` (desugaring requires kind of Type)
> 
>     * `Type|...| @ rc` is desugared `(FromVectorMut.fromVector(rc, $Vector$(...)): Type[_, _, ...])` (desugaring requires the Kind of Type)

I spiritually agree with this. Two questions; how will you handle `Map|...|: ???`? and will you resolve Map using the local scope?

**Comment by JonathanStarup** [comment]
> how will you handle Map|...|: ???

Is this an ascription? It's just desugared directly `((FromVector.fromVector($Vector$(...)): Map[_, _]): ???)`

> will you resolve Map using the local scope?

yes

**Comment by magnus-madsen** [comment]
> Is this an ascription? It's just desugared directly `((FromVector.fromVector($Vector$(...)): Map[_, _]): ???)`

Is that sufficient?

**Comment by mlutze** [comment]
Is the question relating to the fact that map has a different syntax?

```
Map#{"a" => 1, "b" => 2}
```

**Comment by magnus-madsen** [comment]
Just wondering if this will always work out neatly.

**Comment by mlutze** [comment]
See some discussion here: https://www.reddit.com/r/haskell/comments/18ncub/ghc_head_now_has_overloadedlists_as_an_extension/

Note that some of the negatives are in regard to the option of omitting the type, which I am not sure we should allow.

There are other types such as `Nel` which we cannot support directly, unfortunately.

**Comment by JonathanStarup** [comment]
> Just wondering if this will always work out neatly.

It works for List, Set, Vector, Array, and Map at least. Which allow us to remove their ad-hoc constructors

**Comment by mlutze** [comment]
> > Just wondering if this will always work out neatly.
> 
> It works for List, Set, Vector, Array, and Map at least. Which allow us to remove their ad-hoc constructors

Only `Map` if we do extra magic for key -> value mappings

**Comment by magnus-madsen** [comment]
> > > Just wondering if this will always work out neatly.
> > 
> > 
> > It works for List, Set, Vector, Array, and Map at least. Which allow us to remove their ad-hoc constructors
> 
> Only `Map` if we do extra magic for key -> value mappings

Can you expand?

**Comment by mlutze** [comment]
> > > > Just wondering if this will always work out neatly.
> > > 
> > > 
> > > It works for List, Set, Vector, Array, and Map at least. Which allow us to remove their ad-hoc constructors
> > 
> > Only `Map` if we do extra magic for key -> value mappings
> 
> Can you expand?

Other structures have only values. Maps have keys and values. We distinguish between them currently syntactically by doing `Map#{ k => v }`.

This does not immediately lend itself to to a FromVector instance.

**Comment by magnus-madsen** [comment]
Would we then do pairs? Ah- but then keys can overlap

**Comment by JonathanStarup** [comment]
> Only Map if we do extra magic for key -> value mappings

Yeah isn't it just `Map|('a', 1), ('b', 2)|`? and `Map|1, ,2|` will give some instance error

> Would we then do pairs? Ah- but then keys can overlap

These are currently valid `Set#{1, 1, 2}`, `Map#{1 => 1, 1 => 2}` (bonus question: what map do you get?).

This idea isn't trying to save the world. It is just simplifying and generalizing what we have now.

**Comment by mlutze** [comment]
> > Only Map if we do extra magic for key -> value mappings
> 
> Yeah isn't it just `Map|('a', 1), ('b', 2)|`? and `Map|1, ,2|` will give some instance error

Yes, this is just arguably a downgrade in clarity.

> > Would we then do pairs? Ah- but then keys can overlap
> 
> These are currently valid `Set#{1, 1, 2}`, `Map#{1 => 1, 1 => 2}` (bonus question: what map do you get?).
> 
> This idea isn't trying to save the world. It is just simplifying and generalizing what we have now.

Agreed.

**Comment by JonathanStarup** [comment]
> Yes, this is just arguably a downgrade in clarity.

True, but `Map#{'a' => 1, ..}` is also magic syntax not found anywhere else. So maybe the tuples are better for the mental model.
